### ISSUE TITLE
Really Slow Initial Load times for production build. 

### ISSUE THREAD

**Issue by jwrlee88** [jaredpalmer/after.js#345]
### Current Behavior

Hey everyone, big ups to everyone who has been helping to make After.JS version 3 and Razzle awesome. I've been using it pretty successfully, and we are getting ready for an official launch (with some more clean up coming) thanks to help from the likes of @nimaa77 . 

Right now when we deploy the production build live, the site takes about ~30 seconds for the initial load, whereas the content download only really takes 100ms. The fastest we've seen it load is about 22 seconds load and 20 ms for the content download. 

This is a huge discrepancy, and we're wondering if there's a way to shorten the initial load times. :slightly_smiling_face: 

Here is an example of the site loading: 

![loading1](https://user-images.githubusercontent.com/39804005/82703564-26be4600-9c42-11ea-8dca-536d7252d664.png)

![image (2)](https://user-images.githubusercontent.com/39804005/82703610-3e95ca00-9c42-11ea-93b2-81fe6c41e43c.png)

### Expected behavior

We would like the loading time to be shortened so that the user experience for the site improves on initial load. After the initial load, everything is very fast. 

### Additional context

We are using this command in package.json to start the production build: `NODE_ENV=production node build/server.js`

### EDIT: 22nd May 2020 - 20:32 UTC

It seems like `getInitialProps` is being called multiple times per page load. This is the result of my logs showing that its been calling the API multiple times when I just loaded the page:

![api_logs](https://user-images.githubusercontent.com/39804005/82707418-02fefe00-9c4a-11ea-94ca-ccfda66adc78.png)

This might have to do with getInitialProps firing my API call multiple times. Any idea how to fix? 

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| React            | ^16.10.2
| After.js         | ^3.0.2
| Razzle           | ^3.1.0
| Razzle Plugins Scss  | ^ 3.1.0
| Browser          | Google Chrome
| npm/Yarn         | v 1.12.3

**Comment by nimaa77** [comment]
hmm, very weird

can you send me your `routes.js` + the uri that you call

**Comment by nimaa77** [comment]
as a side note
for performance reasons, it's better to serve your static assets on an Nginx or Apache webserver

**Comment by jwrlee88** [comment]
Sure, here is my Routes.js file - you can visit www.scribr.fm <- currently the site is unstable because the server keeps hammering our API, eventually knocking out our server. 

I think there is a correlation between this issue and the issue of getInitialProps firing off API calls via Redux-Saga. 

Let me know if there is any more information I can provide you with, any advice / help would be appreciated.

`Routes.js`: 
```import React from 'react';

import { asyncComponent } from '@jaredpalmer/after';
import LoadingIndicator from "./components/LoadingIndicator";

export default [
  {
    path: '/',
    exact: true,
    component: asyncComponent({
      loader: () => import('./containers/Home/_Home'), // required
      Placeholder: () => <LoadingIndicator/> // this is optional, just returns null by default
    }),
  },
  //*** a bunch more routes not posted for brevity** // 
  {
    path: '',
    exact: true,
    component: asyncComponent({
      loader: () => import('./containers/404/404'),
      Placeholder: () => <LoadingIndicator/>
    })
  }
];
```

**Comment by nimaa77** [comment]
can you create an example repo which reproduces the same problem (I think it's something related to your redux config)

or maybe attaching a debugger could help you to figure it out

https://github.com/jaredpalmer/razzle#npm-start------inspecthostport-or-yarn-start------inspecthostport

nice website btw 🙈

**Comment by jwrlee88** [comment]
> nice website btw :see_no_evil: 

Thank you :smile: 

I am currently getting around the issue of getInitialProps making multiple requests by `await` ing a fetch call, and dispatching success actions to the store in getInitialProps.. Very hacky solution and definitely not redux like ahah

I'm going to attach my `configureStore` and my Document.js and client.js and server.js, I don't think there's anything wrong with the redux configurations but perhaps you could take a look?

configureStore.js:
```
/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware, compose } from 'redux';
import {fromJS} from 'immutable';
import createSagaMiddleware from 'redux-saga'
import createReducer from './reducers';
import rootSaga from './rootSaga'

const sagaMiddleware = createSagaMiddleware();
export default function configureStore(initialState) {
    // Create the store with two middlewares
    // 1. sagaMiddleware: Makes redux-sagas work
    const middlewares = [sagaMiddleware];

    const enhancers = [applyMiddleware(...middlewares)];

    // If Redux DevTools Extension is installed use it, otherwise use Redux compose
    /* eslint-disable no-underscore-dangle, indent */
    const composeEnhancers =
        process.env.NODE_ENV !== 'production' &&
        typeof window === 'object' &&
        window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
            ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
            : compose;
    /* eslint-enable */

    const store = createStore(
        createReducer(),
        fromJS(initialState),
        composeEnhancers(...enhancers),
    );
    // Extensions
    /* istanbul ignore next */
    if (module.hot) {
        module.hot.accept('./reducers', () => {
            const nextRootReducer = require('./reducers').default;
            store.replaceReducer(nextRootReducer);
        });
    }

   sagaMiddleware.run(rootSaga);

    return store;
}

```

server.js: 
```
import React from 'react';
import express from 'express';
import {render} from '@jaredpalmer/after';
import routes from './routes';
import Document from './document';
import configureStore from "./redux/configureStore";

const assets = require(process.env.RAZZLE_ASSETS_MANIFEST);
const chunks = require(process.env.RAZZLE_CHUNKS_MANIFEST); // 👈 import chunks.json

const server = express();
server
  .disable('x-powered-by')
  .use(express.static(process.env.RAZZLE_PUBLIC_DIR))
  .get('/*', async (req, res) => {
    const store = configureStore();
    try {
      const html = await render({
        req,
        res,
        document: Document,
        chunks,
        routes,
        assets,
        // Anything else you add here will be made available
        // within getInitialProps(ctx)
        // e.g a redux store...
        store,
      });
      res.send(html);
    } catch (error) {
        console.error(error);
        res.json({ message: error.message, stack: error.stack });
    }
  });

export default server;
```

document.js:

```
import React from 'react'
import {
  AfterRoot,
  AfterData,
  __AfterContext,
  AfterStyles, AfterScripts, SerializeData
} from '@jaredpalmer/after';
import { Provider } from 'react-redux';
import serialize from 'serialize-javascript';
import { runtimeConfig } from './runtimeConfig'
import './font-awesome'
class Document extends React.Component {

  static async getInitialProps (initialProps) {
    const { renderPage, store } = initialProps

    const page = await renderPage(App => props => (
      <Provider store={store}>
        <App {...props} />
      </Provider>
    ))

    return { ...page }
  }

  render () {
    const { helmet } = this.props
    const htmlAttrs = helmet.htmlAttributes.toComponent()
    const bodyAttrs = helmet.bodyAttributes.toComponent()

    return (
      <html {...htmlAttrs}>
      <head>
        <meta httpEquiv="X-UA-Compatible" content="IE=edge"/>
        <meta charSet="utf-8"/>
        <title>Scribr.FM</title>
        <meta name="viewport" content="width=device-width,initial-scale=1,maximum-scale=1"/>
        {helmet.title.toComponent()}
        {helmet.meta.toComponent()}
        {helmet.link.toComponent()}
        </script>
        <AfterStyles/>
      </head>
      <body {...bodyAttrs}>
        <noscript dangerouslySetInnerHTML={{
          __html:
            `<iframe src="https://www.googletagmanager.com/ns.html?id=GTM-N53HDZ4"
                              height="0" width="0" style="display:none;visibility:hidden">
                </iframe>`
        }}>
        </noscript>
        <AfterRoot/>
        <AfterData/>
        <AfterScripts/>
        <ReduxData/>
        <LoadEnvVars/>
      </body>
      </html>
    )
  }
}

function LoadEnvVars() {
  return (
    <script
      dangerouslySetInnerHTML={{
        __html: `window.env = ${serialize(runtimeConfig)}`,
      }}
    />
  )
}

function ReduxData() {
  const { store } = React.useContext(__AfterContext);
  return <SerializeData name="preloaded_state" data={store.getState()} />;
}

export default Document
```

client.js: 
```
import React from 'react'
import { hydrate } from 'react-dom'
import { BrowserRouter } from 'react-router-dom'
import {ensureReady, After, getSerializedData} from '@jaredpalmer/after'
import { Provider } from 'react-redux'
import configureStore from './redux/configureStore'
import './font-awesome'
import routes from './routes'
import './client.scss'
import AppContainer from './AppContainer'

const preloadedState = getSerializedData('preloaded_state');
const store = configureStore(preloadedState);

function renderApp () {
  ensureReady(routes).then(data =>
    hydrate(
      <BrowserRouter>
        <Provider store={store}>
            <After data={data} routes={routes} store={store} />
        </Provider>
      </BrowserRouter>,
      document.getElementById('root')
    )
  )
}

renderApp()

if (module.hot) {
  module.hot.accept('./routes', renderApp)
}

```

**Comment by nimaa77** [comment]
the redux configuration is correct.

I'm not a saga pro but this may help

```js
  await store.sagaTask.toPromise()
```

https://github.com/zeit/next.js/blob/canary/examples/with-redux-saga/pages/index.js#L28

**Comment by jwrlee88** [comment]
So the issue with the slow loading times was really because our Redux Saga configs were off - we were able to get around the issue by doing what I described in a previous comment: 

> I am currently getting around the issue of getInitialProps making multiple requests by await ing a fetch call, and dispatching success actions to the store in getInitialProps.. Very hacky solution and definitely not redux like ahah

The next.js example that Nimaa linked in [the previous comment](https://github.com/jaredpalmer/after.js/issues/345#issuecomment-634083920) helped but I still saw the API getting hit repeatedly. I think its due to the fact that the next.js example uses `next-redux-saga` and I don't. I'm not really sure if that library is compatible with After.js, perhaps someone else might be able to play around with it and make it work with After.js? Not sure. 

Anyways, thank you for the help @nimaa77 , I will be closing this issue now as the original problem is now fixed.
